### PR TITLE
Live Range Analysis and Domination Analysis use Static Information 

### PR DESCRIPTION
**Pull request by calebmkim**
I already implemented sharing across static pars using the `static` annotation, so all I had to do for that was using the static control. 

For adding dominators across par blocks, here is what we do: 
1) We first build a domination map normally w/o using any static par information
2) Then, we do a subsequent analysis (in `static_par_domination.rs`) that adds dominators across static par blocks. For this analysis, we map nodes (i.e., enables and if condition ports) to the intervals that they first execute. There are a few things to note about this:
- We make a distinction between nodes that _must_ execute and nodes that _may_ execute. Essentially, any nodes inside an `if` branch _may_ execute, and everything else _must_ execute. "may execute" nodes can be dominated but they should not dominate anything, whereas "must execute" nodes can dominate other nodes and be dominated. 
- We are only interested in the first execution of nodes, so unlike with Live Range Analysis, we don't need a `Vec` of intervals that the group is live, we only need the first interval it executes. 
After mapping these nodes, we do a very simple algorithm, that just looks at the execution interval for each possible dominator of each node, and adds entries to the dominator map when appropriate. 

**Another Note** 
I changed the timing intervals on the static par analysis for live ranges to start at 0 and be open/exclusive at the end, to align with the `%` syntax we have already created.

### CODE REVIEW

**Comment by rachitnigam** [comment]
Looking good! Some high-level questions:

> We are only interested in the first execution of nodes, so unlike with Live Range Analysis, we don't need a Vec of intervals that the group is live, we only need the first interval it executes.

Can you talk a bit more about this? What does "first execution of nodes" mean? Are you talking about the first time a group executes?

> After mapping these nodes, we do a very simple algorithm, that just looks at the execution interval for each possible dominator of each node, and adds entries to the dominator map when appropriate

Does this have the effect of making the domination analysis more precise? How so?

(Also, feel free to merge. Not sure if I'll have useful low-level comments)

File: calyx-opt/src/analysis/mod.rs [hunk 1]
```diff
@@ -17,6 +17,7 @@ pub mod reaching_defns;
 mod read_write_set;
 mod schedule_conflicts;
 mod share_set;
+mod static_par_domination;
```
Comment (rachitnigam) [inline calyx-opt/src/analysis/mod.rs:20]: Is there ever a case we'd want to run `static_par_domination` separately from the full domination pass?

File: calyx-opt/src/analysis/static_par_domination.rs [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+            let may_execute_enables =
+                match self.node_maybe_timing_map.get(par_id) {
+                    Some(mapping) => mapping,
+                    None => &empty_map,
+                };
+            // Very simple/naive algorithm
+            // for each "must" execute nodes, it can either dominate:
+            // one of the "may" execute nodes, or dominate another "must" execute node
+            // "may" execute nodes cannot dominate anybody
+            for (node_id1, (_, end1)) in node_interval_mapping.clone() {
```
Comment (rachitnigam) [inline calyx-opt/src/analysis/static_par_domination.rs:117]: Why is this `clone()` necessary? Can you iterate over the set and only clone the underlying numbers when needed?

File: calyx-opt/src/analysis/static_par_domination.rs [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+        let (par_id, cur_clock) = cur_state;
+        // if we are guaranteed execution, then can update self.node_timing_map
+        // otherwise we must updateself.node_maybe_timing_map
+        let enable_mappings = if guaranteed_execute {
+            self.node_timing_map.entry(par_id).or_default()
+        } else {
+            self.node_maybe_timing_map.entry(par_id).or_default()
+        };
+        // maps enable ids -> clock cycles that they're live in
+        match enable_mappings.get(&id) {
```
Comment (rachitnigam) [inline calyx-opt/src/analysis/static_par_domination.rs:162]: The `match` is uneccesary. Turn this into:
```
if enable_mappings.get(&id).is_none() {
 ...
 }
```

File: calyx-opt/src/analysis/static_par_timing.rs [hunk 4]
```diff
@@ -14,7 +14,7 @@ type ThreadTimingMap = HashMap<u64, CellTimingMap>;
 #[derive(Default)]
 pub struct StaticParTiming {
     /// Map from par block ids to cell_timing_maps
-    cell_map: HashMap<u64, ThreadTimingMap>,
+    pub cell_map: HashMap<u64, ThreadTimingMap>,
```
Comment (rachitnigam) [inline calyx-opt/src/analysis/static_par_timing.rs:17]: Why is this public? Should this be accessible to consumers of this pass? If so, they should not be able to mutate the field in which case we should provide an accessor method.

**Comment by rachitnigam** [comment]
Also, does this make the component-sharing pass more precise as well?

**Comment by calebmkim** [comment]
> Can you talk a bit more about this? What does "first execution of nodes" mean? Are you talking about the first time a group executes?

Yeah, it's the first time the group executes (of an `if` port is read from). 

> Does this have the effect of making the domination analysis more precise? How so?

Yes, since it accounts for domination across par blocks. e.g: 
```
static par {
    static seq {A<2>; B<2>} 
    C<2>; 
} 
```
We can now infer that `C` dominates `B`. `static-par-domination` does an analysis for each `static` par block to determine when, relative to the start of the par block, each of the groups executes for the first time. 
The test cases demonstrate this (although some of them got a bit complicated). 

> Also, does this make the component-sharing pass more precise as well?

Yes. There's a very similar analysis in `static-par-timing`

**Comment by rachitnigam** [comment]
Got it! I think this change makes sense. Long term, we should extract the commonality between the two analysis into a more generic analysis that might be applicable for more cases.

**Pull request merged**